File: src/executor.ts
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import type { SubscriptionConfig, SmeeEventData, ExecutionResult } from './types.js';

const execFileAsync = promisify(execFile);

/**
 * 将 Smee 事件数据和基础提示词合并成完整的提示词
 */
function buildPrompt(basePrompt: string, event: SmeeEventData): string {
  const bodyStr =
    typeof event.body === 'string'
      ? event.body
      : JSON.stringify(event.body, null, 2);

  const headersStr = JSON.stringify(event.headers, null, 2);
  const queryStr =
    Object.keys(event.query).length > 0
      ? JSON.stringify(event.query, null, 2)
      : '（无）';

  return [
    basePrompt.trim(),
    '',
    '---',
    '以下是通过 Webhook 接收到的事件信息，请结合上述指令处理：',
    '',
    '## 请求头（Headers）',
    '```json',
    headersStr,
    '```',
    '',
    '## 查询参数（Query Parameters）',
    '```json',
    queryStr,
    '```',
    '',
    '## 请求体（Body）',
    '```',
    bodyStr,
    '```',
    '',
    `> 事件接收时间：${new Date(event.timestamp).toISOString()}`,
  ].join('\n');
}

/**
 * 调用 Claude CLI 执行提示词
 * 使用 claude -p "<prompt>" --dangerously-skip-permissions
 */
async function runClaude(
  prompt: string,
  workspace: string
): Promise<{ output: string; success: boolean; error?: string }> {
  try {
    const { stdout, stderr } = await execFileAsync(
      'claude',
      ['-p', prompt, '--dangerously-skip-permissions'],
      {
        cwd: workspace,
        // 不设置超时，让 claude 自行完成
        maxBuffer: 100 * 1024 * 1024, // 100MB 输出缓冲
      }
    );

    const output = stdout + (stderr ? `\n[stderr]\n${stderr}` : '');
    return { output, success: true };
  } catch (err: unknown) {
    const error = err as NodeJS.ErrnoException & {
      stdout?: string;
      stderr?: string;
    };
    return {
      output: [error.stdout ?? '', error.stderr ?? ''].filter(Boolean).join('\n'),
      success: false,
      error: error.message,
    };
  }
}

/**
 * 执行器主函数：将事件组装为提示词并调用 Claude
 */
export async function executeWithClaude(
  subscription: SubscriptionConfig,
  defaultWorkspace: string,
  event: SmeeEventData
): Promise<ExecutionResult> {
  const workspace = subscription.workspace ?? defaultWorkspace;
  const prompt = buildPrompt(subscription.basePrompt, event);
  const startedAt = new Date();

  console.log(
    `\n🤖 [${subscription.name}] 开始调用 Claude（工作区: ${workspace}）...`
  );

  const { output, success, error } = await runClaude(prompt, workspace);
  const finishedAt = new Date();

  return {
    subscriptionName: subscription.name,
    success,
    output,
    error,
    startedAt,
    finishedAt,
    prompt,
  };
}

File: rust/src/executor.rs
use anyhow::Result;
use chrono::Utc;
use tokio::process::Command;

use crate::types::{ExecutionResult, SmeeEventData, SubscriptionConfig};

fn build_prompt(base_prompt: &str, event: &SmeeEventData) -> String {
    let body_str = match &event.body {
        serde_json::Value::String(s) => s.clone(),
        v => serde_json::to_string_pretty(v).unwrap_or_default(),
    };

    let headers_str = serde_json::to_string_pretty(&event.headers).unwrap_or_default();

    let query_str = if event.query.is_empty() {
        "（无）".to_string()
    } else {
        serde_json::to_string_pretty(&event.query).unwrap_or_default()
    };

    let timestamp = chrono::DateTime::from_timestamp_millis(event.timestamp as i64)
        .map(|dt| dt.to_rfc3339())
        .unwrap_or_default();

    format!(
        "{base_prompt}\n\n---\n以下是通过 Webhook 接收到的事件信息，请结合上述指令处理：\n\n## 请求头（Headers）\n```json\n{headers_str}\n```\n\n## 查询参数（Query Parameters）\n```json\n{query_str}\n```\n\n## 请求体（Body）\n```\n{body_str}\n```\n\n> 事件接收时间：{timestamp}",
        base_prompt = base_prompt.trim(),
    )
}

async fn run_claude(prompt: &str, workspace: &str) -> (String, bool, Option<String>) {
    let result = Command::new("claude")
        .args(["-p", prompt, "--dangerously-skip-permissions"])
        .current_dir(workspace)
        .output()
        .await;

    match result {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout).into_owned();
            let stderr = String::from_utf8_lossy(&output.stderr).into_owned();
            let combined = if stderr.is_empty() {
                stdout
            } else {
                format!("{stdout}\n[stderr]\n{stderr}")
            };

            if output.status.success() {
                (combined, true, None)
            } else {
                let err = format!(
                    "进程退出码: {}",
                    output.status.code().unwrap_or(-1)
                );
                (combined, false, Some(err))
            }
        }
        Err(e) => (String::new(), false, Some(e.to_string())),
    }
}

pub async fn execute_with_claude(
    subscription: &SubscriptionConfig,
    default_workspace: &str,
    event: &SmeeEventData,
) -> Result<ExecutionResult> {
    let workspace = subscription
        .workspace
        .as_deref()
        .unwrap_or(default_workspace);
    let prompt = build_prompt(&subscription.base_prompt, event);
    let started_at = Utc::now();

    println!(
        "\n🤖 [{}] 开始调用 Claude（工作区: {}）...",
        subscription.name, workspace
    );

    let (output, success, error) = run_claude(&prompt, workspace).await;
    let finished_at = Utc::now();

    Ok(ExecutionResult {
        subscription_name: subscription.name.clone(),
        success,
        output,
        error,
        started_at,
        finished_at,
        prompt,
    })
}
